scalableminds/webknossos
# Import NML in frontend

This PR adds an NML import to the frontend. I've pretty much duplicated the backend functionality in this regard. There is error checking for malformed NMLs and a check whether all trees are actually connected. The parser uses some default values for missing optional properties, I've copied those from the backend as well.
I've also added unit tests that test the NML export in conjunction with the NML import, error handling and NML naming.

### Mailable description of changes:
 - It is now possible to import an NML into an existing tracing, from within the tracing view. Select `Import NML` from the dropdown in the trees tab. The trees from the NML will be added to the current tracing - trees and nodes will get new ids.

### Steps to test:
- Run tests.
- Download an NML (using backend/frontend download) and import it again. The trees should be added to the tracing, but there should be no duplicate ids - comments and branchpoints should have no id conflicts as well and still belong to the same nodes.
- Try to upload malformed NMLs with duplicate node/tree ids, incorrect edges, disconnected trees, whatever you can think of. There should be an error Toast, explaining the error during import.

### Issues:
- fixes #2094

------
